src/color.rs

```diff
@@ -1060,87 +1149,149 @@ fn parse_rgb<'i, 't, P>(
 where
     P: ColorParser<'i>,
 {
-    // Either integers or percentages, but all the same type.
-    // https://drafts.csswg.org/css-color/#rgb-functions
-    let (red, is_number) = match color_parser.parse_number_or_percentage(arguments)? {
-        NumberOrPercentage::Number { value } => (clamp_floor_256_f32(value), true),
-        NumberOrPercentage::Percentage { unit_value } => (clamp_unit_f32(unit_value), false),
+    let location = arguments.current_source_location();
+
+    // rgb(..) and rgba(..) allows both legacy and new syntax, so try legacy
+    // first and if that fails, try the newer syntax. The newer syntax still
+    // does not support "none".
+    let start = arguments.state();
+    let (red, green, blue, alpha) = match parse_legacy_components(
+        color_parser,
+        arguments,
+        P::parse_number_or_percentage,
+        P::parse_number_or_percentage,
+        P::parse_number_or_percentage,
+    ) {
+        Ok(c) => c,
+        Err(_) => {
+            arguments.reset(&start);
+            parse_components(
+                color_parser,
+                arguments,
+                P::parse_number_or_percentage,
+                P::parse_number_or_percentage,
+                P::parse_number_or_percentage,
+            )?
+        }
     };
 
-    let uses_commas = arguments.try_parse(|i| i.expect_comma()).is_ok();
+    // Components are only valid if they are all numbers or all percentages.
 
-    let green;
-    let blue;
-    if is_number {
-        green = clamp_floor_256_f32(color_parser.parse_number(arguments)?);
-        if uses_commas {
-            arguments.expect_comma()?;
-        }
-        blue = clamp_floor_256_f32(color_parser.parse_number(arguments)?);
-    } else {
-        green = clamp_unit_f32(color_parser.parse_percentage(arguments)?);
-        if uses_commas {
-            arguments.expect_comma()?;
+    // 0 = none
+    // 1 = number
+    // 2 = percentage
+    let mut types = vec![];
```

Comment:
I'm pretty sure having an extra heap-allocation here is not acceptable... Why can't we keep the rgb code as it was? It's the hottest one by far, and it doesn't seem it gives us much factoring out if we then need to filter the types manually. Can we split the component parsing changes from the `Hsl` / `Hwb` changes?

Reply:
Parsing changes are now in a separate PR.